cargo test duplicate error message 3 times

<!-- Thanks for filing a 🐛 bug report 😄! -->

**Problem**

Im using workspace with three projects frontend(wasm), backend(native) and common-api (shared between frontend and backend).

```cargo test -p backend``` deadly slow and then pumping same error messages 3 times.

Output of `cargo version`:

```
$ cargo --version
cargo 1.33.0 (f099fe94b 2019-02-12)

ubuntu 18
$ uname -r
4.18.0-16-generic
$ rustc --version
rustc 1.33.0 (2aa4c46cf 2019-02-28)
```

```
cargo test -v -p ast-backend ${ARGS} -- --test-threads 1
       Fresh semver-parser v0.7.0
       Fresh autocfg v0.1.2
       Fresh rand_core v0.4.0
       Fresh void v1.0.2
       Fresh cfg-if v0.1.6
       Fresh stable_deref_trait v1.1.1
       Fresh nodrop v0.1.13
       Fresh unicode-xid v0.1.0
       Fresh lazy_static v1.2.0
       Fresh cc v1.0.29
       Fresh version_check v0.1.5
       Fresh memoffset v0.2.1
       Fresh scopeguard v0.3.3
       Fresh matches v0.1.8
       Fresh futures v0.1.25
       Fresh encoding_index_tests v0.1.4
       Fresh lazycell v1.2.1
       Fresh slab v0.4.2
       Fresh rustc-demangle v0.1.13
       Fresh percent-encoding v1.0.1
       Fresh pkg-config v0.3.14
       Fresh arc-swap v0.3.7
       Fresh quick-error v1.2.2
       Fresh linked-hash-map v0.4.2
       Fresh build_const v0.2.1
       Fresh gcc v0.3.55
       Fresh siphasher v0.2.3
       Fresh antidote v1.0.0
       Fresh itoa v0.4.3
       Fresh fnv v1.0.6
       Fresh bitflags v1.0.4
       Fresh untrusted v0.6.2
       Fresh ucd-util v0.1.3
       Fresh safemem v0.3.0
       Fresh dtoa v0.4.3
       Fresh half v1.3.0
       Fresh utf8-ranges v1.0.2
       Fresh string v0.1.3
       Fresh remove_dir_all v0.5.1
       Fresh byteorder v0.5.3
       Fresh rustc-serialize v0.3.24
       Fresh indexmap v1.0.2
       Fresh language-tags v0.2.2
       Fresh hex v0.3.2
       Fresh fishblow v0.1.0 (/home/dd/work/ast_store/deps/fishblow)
       Fresh sha1 v0.6.0
       Fresh termcolor v1.0.4
       Fresh etst v0.1.0 (/home/dd/work/ast_store/deps/etst)
       Fresh either v1.5.0
       Fresh same-file v1.0.4
       Fresh semver v0.9.0
       Fresh rand_core v0.3.1
       Fresh rand_jitter v0.1.3
       Fresh unreachable v1.0.0
       Fresh crossbeam-utils v0.2.2
       Fresh log v0.4.6
       Fresh owning_ref v0.4.0
       Fresh crossbeam-utils v0.6.5
       Fresh thread_local v0.3.6
       Fresh unicode-bidi v0.3.4
       Fresh tokio-sync v0.1.1
       Fresh tower-service v0.1.0
       Fresh encoding-index-japanese v1.20141219.5
       Fresh encoding-index-singlebyte v1.20141219.5
       Fresh encoding-index-tradchinese v1.20141219.5
       Fresh encoding-index-korean v1.20141219.5
       Fresh encoding-index-simpchinese v1.20141219.5
       Fresh humantime v1.2.0
       Fresh lru-cache v0.1.1
       Fresh scheduled-thread-pool v0.2.0
       Fresh regex-syntax v0.6.5
       Fresh riff-wave v0.1.2
       Fresh walkdir v2.2.7
       Fresh rustc_version v0.2.3
       Fresh libc v0.2.48
       Fresh rand_hc v0.1.0
       Fresh rand_xorshift v0.1.1
       Fresh rand_isaac v0.1.1
       Fresh proc-macro2 v0.4.27
       Fresh arrayvec v0.4.10
       Fresh smallvec v0.6.8
       Fresh byteorder v1.3.1
       Fresh lock_api v0.1.5
       Fresh tokio-executor v0.1.6
       Fresh encoding v0.2.33
       Fresh num-traits v0.2.6
       Fresh ryu v0.2.7
       Fresh serde v1.0.87
       Fresh crc32fast v1.1.2
       Fresh pq-sys v0.4.6
       Fresh r2d2 v0.8.3
       Fresh httparse v1.3.3
       Fresh rand_os v0.1.2
       Fresh iovec v0.1.2
       Fresh net2 v0.2.33
       Fresh num_cpus v1.9.0
       Fresh rand v0.5.6
       Fresh socket2 v0.3.8
       Fresh hostname v0.1.5
       Fresh signal-hook v0.1.7
       Fresh rand v0.4.6
       Fresh time v0.1.42
       Fresh memchr v2.1.3
       Fresh filetime v0.2.4
       Fresh xattr v0.2.2
       Fresh memmap v0.6.2
       Fresh atty v0.2.11
       Fresh listenfd v0.3.3
       Fresh rand_chacha v0.1.1
       Fresh quote v0.6.11
       Fresh crossbeam-epoch v0.3.1
       Fresh crossbeam-epoch v0.7.1
       Fresh crossbeam-channel v0.3.8
       Fresh unicode-normalization v0.1.8
       Fresh base64 v0.9.3
       Fresh base64 v0.10.1
       Fresh tokio-timer v0.2.10
       Fresh tokio-current-thread v0.1.4
       Fresh backtrace-sys v0.1.28
       Fresh ring v0.13.5
       Fresh brotli-sys v0.3.2
       Fresh miniz-sys v0.1.11
       Fresh unicase v1.4.2
       Fresh unicase v2.2.0
       Fresh num-integer v0.1.39
       Fresh libsqlite3-sys v0.9.3
       Fresh serde_json v1.0.38
       Fresh serde_cbor v0.9.0
       Fresh crc v1.8.1
       Fresh nom v4.2.0
       Fresh bytes v0.4.11
       Fresh mio v0.6.16
       Fresh futures-cpupool v0.1.8
       Fresh resolv-conf v0.6.2
       Fresh rand v0.3.23
       Fresh aho-corasick v0.6.9
       Fresh tar v0.4.20
       Fresh syn v0.15.26
       Fresh crossbeam-deque v0.2.0
       Fresh crossbeam-deque v0.6.3
       Fresh idna v0.1.5
       Fresh backtrace v0.3.13
       Fresh brotli2 v0.3.2
       Fresh flate2 v1.0.6
       Fresh phf_shared v0.7.24
       Fresh mime v0.3.13
       Fresh chrono v0.4.6
       Fresh rusqlite v0.13.0
       Fresh rand_pcg v0.1.1
       Fresh tokio-io v0.1.11
       Fresh http v0.1.15
       Fresh mio-uds v0.6.7
       Fresh rust-crypto v0.2.36
       Fresh regex v1.1.0
       Fresh synstructure v0.10.1
       Fresh actix_derive v0.3.2
       Fresh v_escape_derive v0.2.1
       Fresh diesel_derives v1.4.0
       Fresh serde_derive v1.0.87
       Fresh rayon-core v1.4.1
       Fresh url v1.7.2
       Fresh phf v0.7.24
       Fresh rand v0.6.5
       Fresh tokio-codec v0.1.1
       Fresh h2 v0.1.16
       Fresh astindex2 v0.1.0 (/home/dd/work/ast_store/deps/astindex2)
       Fresh asfile v0.1.0 (/home/dd/work/ast_store/asfile)
       Fresh env_logger v0.6.0
       Fresh failure_derive v0.1.5
       Fresh v_escape v0.3.2
       Fresh jsonwebtoken v5.0.1
       Fresh rayon v1.0.3
       Fresh cookie v0.11.0
       Fresh serde_urlencoded v0.5.4
       Fresh parking_lot_core v0.4.0
       Fresh phf_generator v0.7.24
       Fresh uuid v0.7.2
       Fresh tempfile v3.0.6
       Fresh failure v0.1.5
       Fresh v_htmlescape v0.3.2
       Fresh parking_lot v0.7.1
       Fresh phf_codegen v0.7.24
       Fresh diesel v1.4.1
       Fresh mp3-duration v0.1.4
       Fresh dotenv v0.13.0
       Fresh astu-fw v0.1.0 (/home/dd/work/ast_store/astu-fw)
       Fresh crossbeam v0.6.0
       Fresh tokio-reactor v0.1.8
       Fresh ast-store v0.1.0 (/home/dd/work/ast_store)
       Fresh asc-container v0.1.0 (/home/dd/work/ast_store/asc-container)
       Fresh tokio-threadpool v0.1.11
       Fresh tokio-uds v0.2.5
       Fresh tokio-tcp v0.1.3
       Fresh tokio-udp v0.1.3
       Fresh tokio-signal v0.2.7
       Fresh tokio-fs v0.1.5
       Fresh trust-dns-proto v0.6.3
       Fresh trust-dns-proto v0.5.0
       Fresh mime_guess v2.0.0-alpha.6
       Fresh tokio v0.1.15
       Fresh trust-dns-resolver v0.10.3
       Fresh actix v0.7.9
       Fresh actix-net v0.2.6
       Fresh actix-web v0.7.18
   Compiling ast-backend v0.1.0 (/home/dd/work/ast_store/backend)
     Running `rustc --edition=2018 --crate-name ast_backend backend/src/lib.rs --color always --crate-type lib --emit=dep-info,link -C opt-level=1 -C debuginfo=2 -C debug-assertions=on -C metadata=f83015185db9fbc0 -C extra-filename=-f83015185db9fbc0 --out-dir /home/dd/work/ast_store/target/debug/deps -C incremental=/home/dd/work/ast_store/target/debug/incremental -L dependency=/home/dd/work/ast_store/target/debug/deps --extern actix=/home/dd/work/ast_store/target/debug/deps/libactix-33f6a9c015b4c03c.rlib --extern actix_web=/home/dd/work/ast_store/target/debug/deps/libactix_web-a11a56b0a62f4234.rlib --extern asc_container=/home/dd/work/ast_store/target/debug/deps/libasc_container-06f10de12fae04ac.rlib --extern asfile=/home/dd/work/ast_store/target/debug/deps/libasfile-4547961b3775e29e.rlib --extern ast_store=/home/dd/work/ast_store/target/debug/deps/libast_store-c5777bddc6b7a1ab.rlib --extern astindex2=/home/dd/work/ast_store/target/debug/deps/libastindex2-d64d1b076e5c74e6.rlib --extern astu_fw=/home/dd/work/ast_store/target/debug/deps/libastu_fw-43532442dc812bda.rlib --extern base64=/home/dd/work/ast_store/target/debug/deps/libbase64-f864812a9525c302.rlib --extern bytes=/home/dd/work/ast_store/target/debug/deps/libbytes-2e777982cd476fc8.rlib --extern chrono=/home/dd/work/ast_store/target/debug/deps/libchrono-31d6e1ac99b55645.rlib --extern diesel=/home/dd/work/ast_store/target/debug/deps/libdiesel-5eae397717655771.rlib --extern dotenv=/home/dd/work/ast_store/target/debug/deps/libdotenv-1fd097d50a0cbea9.rlib --extern env_logger=/home/dd/work/ast_store/target/debug/deps/libenv_logger-cf98bd5127572178.rlib --extern failure=/home/dd/work/ast_store/target/debug/deps/libfailure-da7463ce07672761.rlib --extern futures=/home/dd/work/ast_store/target/debug/deps/libfutures-fe721e4f842dc564.rlib --extern hex=/home/dd/work/ast_store/target/debug/deps/libhex-801f14eedb60222c.rlib --extern jsonwebtoken=/home/dd/work/ast_store/target/debug/deps/libjsonwebtoken-00f2e8867e3b1c8f.rlib --extern listenfd=/home/dd/work/ast_store/target/debug/deps/liblistenfd-b75707ba07f921d4.rlib --extern log=/home/dd/work/ast_store/target/debug/deps/liblog-fe56020e31c924f7.rlib --extern r2d2=/home/dd/work/ast_store/target/debug/deps/libr2d2-73c739306789f716.rlib --extern rand=/home/dd/work/ast_store/target/debug/deps/librand-e06e57be902b350f.rlib --extern crypto=/home/dd/work/ast_store/target/debug/deps/libcrypto-84a7de1b9fd676db.rlib --extern serde=/home/dd/work/ast_store/target/debug/deps/libserde-3127abc1b61429fa.rlib --extern serde_cbor=/home/dd/work/ast_store/target/debug/deps/libserde_cbor-edda23d0240421b5.rlib --extern serde_derive=/home/dd/work/ast_store/target/debug/deps/libserde_derive-efef5740544450c3.so --extern serde_json=/home/dd/work/ast_store/target/debug/deps/libserde_json-c1cd94b393a59f47.rlib --extern tempfile=/home/dd/work/ast_store/target/debug/deps/libtempfile-ba69c403ad632a75.rlib --extern tokio=/home/dd/work/ast_store/target/debug/deps/libtokio-cdda36ea935a2ce3.rlib --extern uuid=/home/dd/work/ast_store/target/debug/deps/libuuid-476d7cf0d1a23342.rlib -L native=/home/dd/work/ast_store/target/debug/build/backtrace-sys-ab2de583a5a63dbe/out -L native=/home/dd/work/ast_store/target/debug/build/brotli-sys-95a10507b4d53fe6/out -L native=/home/dd/work/ast_store/target/debug/build/ring-aa181c5e8c6d15dd/out -L native=/home/dd/work/ast_store/target/debug/build/miniz-sys-cb9ad76ca80f6a05/out -L native=/home/dd/work/ast_store/target/debug/build/asc-container-12a918cca80da199/out -L native=/home/dd/work/ast_store/target/debug/build/rust-crypto-f98b19eda631acd4/out -L native=/usr/lib/x86_64-linux-gnu -L native=/home/dd/work/ast_store/target/debug/build/libsqlite3-sys-14ba8e47f5445efc/out`
     Running `rustc --edition=2018 --crate-name ast_backend backend/src/lib.rs --color always --crate-type lib --emit=dep-info,link -C opt-level=1 -C panic=unwind -C debuginfo=2 -C debug-assertions=on -C metadata=235be9f8c403ef5f -C extra-filename=-235be9f8c403ef5f --out-dir /home/dd/work/ast_store/target/debug/deps -C incremental=/home/dd/work/ast_store/target/debug/incremental -L dependency=/home/dd/work/ast_store/target/debug/deps --extern actix=/home/dd/work/ast_store/target/debug/deps/libactix-e6e6ec5989c58f08.rlib --extern actix_web=/home/dd/work/ast_store/target/debug/deps/libactix_web-0c918bea4ce6e184.rlib --extern asc_container=/home/dd/work/ast_store/target/debug/deps/libasc_container-79248c028c0abaa6.rlib --extern asfile=/home/dd/work/ast_store/target/debug/deps/libasfile-501d87742c18de63.rlib --extern ast_store=/home/dd/work/ast_store/target/debug/deps/libast_store-e563fff47db84ca0.rlib --extern astindex2=/home/dd/work/ast_store/target/debug/deps/libastindex2-645dc7c64d98dc4b.rlib --extern astu_fw=/home/dd/work/ast_store/target/debug/deps/libastu_fw-eedd0a1b56692ddc.rlib --extern base64=/home/dd/work/ast_store/target/debug/deps/libbase64-e6a7e4ff2fe730df.rlib --extern bytes=/home/dd/work/ast_store/target/debug/deps/libbytes-e8b41f704e29044f.rlib --extern chrono=/home/dd/work/ast_store/target/debug/deps/libchrono-c9d85afdd546881d.rlib --extern diesel=/home/dd/work/ast_store/target/debug/deps/libdiesel-a097740f167c8b3a.rlib --extern dotenv=/home/dd/work/ast_store/target/debug/deps/libdotenv-7f00021dad648ea9.rlib --extern env_logger=/home/dd/work/ast_store/target/debug/deps/libenv_logger-97aaeafa161d4e8c.rlib --extern failure=/home/dd/work/ast_store/target/debug/deps/libfailure-e4fe55d2119b08e5.rlib --extern futures=/home/dd/work/ast_store/target/debug/deps/libfutures-692e72e42ffcb5bf.rlib --extern hex=/home/dd/work/ast_store/target/debug/deps/libhex-b793ef693e61c776.rlib --extern jsonwebtoken=/home/dd/work/ast_store/target/debug/deps/libjsonwebtoken-209fe79e69c28919.rlib --extern listenfd=/home/dd/work/ast_store/target/debug/deps/liblistenfd-c51dae87f5f318a2.rlib --extern log=/home/dd/work/ast_store/target/debug/deps/liblog-01c20663ef7f3d1f.rlib --extern r2d2=/home/dd/work/ast_store/target/debug/deps/libr2d2-0028d1814f157a08.rlib --extern rand=/home/dd/work/ast_store/target/debug/deps/librand-a37ae1bb175fe8d8.rlib --extern crypto=/home/dd/work/ast_store/target/debug/deps/libcrypto-4ef6247b79c391bc.rlib --extern serde=/home/dd/work/ast_store/target/debug/deps/libserde-aeb0ce1d2c705269.rlib --extern serde_cbor=/home/dd/work/ast_store/target/debug/deps/libserde_cbor-466e451a04b77cb2.rlib --extern serde_derive=/home/dd/work/ast_store/target/debug/deps/libserde_derive-efef5740544450c3.so --extern serde_json=/home/dd/work/ast_store/target/debug/deps/libserde_json-0454996f9a0511c1.rlib --extern tempfile=/home/dd/work/ast_store/target/debug/deps/libtempfile-a5cc25e0c0860958.rlib --extern tokio=/home/dd/work/ast_store/target/debug/deps/libtokio-1eecfad9964ed632.rlib --extern uuid=/home/dd/work/ast_store/target/debug/deps/libuuid-5d3f8589b71f0f58.rlib -L native=/home/dd/work/ast_store/target/debug/build/backtrace-sys-ab2de583a5a63dbe/out -L native=/home/dd/work/ast_store/target/debug/build/brotli-sys-95a10507b4d53fe6/out -L native=/home/dd/work/ast_store/target/debug/build/ring-aa181c5e8c6d15dd/out -L native=/home/dd/work/ast_store/target/debug/build/miniz-sys-cb9ad76ca80f6a05/out -L native=/home/dd/work/ast_store/target/debug/build/asc-container-12a918cca80da199/out -L native=/home/dd/work/ast_store/target/debug/build/rust-crypto-f98b19eda631acd4/out -L native=/usr/lib/x86_64-linux-gnu -L native=/home/dd/work/ast_store/target/debug/build/libsqlite3-sys-14ba8e47f5445efc/out`
     Running `rustc --edition=2018 --crate-name ast_backend backend/src/lib.rs --color always --emit=dep-info,link -C opt-level=1 -C debuginfo=2 -C debug-assertions=on --test -C metadata=b552aafba69820d4 -C extra-filename=-b552aafba69820d4 --out-dir /home/dd/work/ast_store/target/debug/deps -C incremental=/home/dd/work/ast_store/target/debug/incremental -L dependency=/home/dd/work/ast_store/target/debug/deps --extern actix=/home/dd/work/ast_store/target/debug/deps/libactix-33f6a9c015b4c03c.rlib --extern actix_web=/home/dd/work/ast_store/target/debug/deps/libactix_web-a11a56b0a62f4234.rlib --extern asc_container=/home/dd/work/ast_store/target/debug/deps/libasc_container-06f10de12fae04ac.rlib --extern asfile=/home/dd/work/ast_store/target/debug/deps/libasfile-4547961b3775e29e.rlib --extern ast_store=/home/dd/work/ast_store/target/debug/deps/libast_store-c5777bddc6b7a1ab.rlib --extern astindex2=/home/dd/work/ast_store/target/debug/deps/libastindex2-d64d1b076e5c74e6.rlib --extern astu_fw=/home/dd/work/ast_store/target/debug/deps/libastu_fw-43532442dc812bda.rlib --extern base64=/home/dd/work/ast_store/target/debug/deps/libbase64-f864812a9525c302.rlib --extern bytes=/home/dd/work/ast_store/target/debug/deps/libbytes-2e777982cd476fc8.rlib --extern chrono=/home/dd/work/ast_store/target/debug/deps/libchrono-31d6e1ac99b55645.rlib --extern diesel=/home/dd/work/ast_store/target/debug/deps/libdiesel-5eae397717655771.rlib --extern dotenv=/home/dd/work/ast_store/target/debug/deps/libdotenv-1fd097d50a0cbea9.rlib --extern env_logger=/home/dd/work/ast_store/target/debug/deps/libenv_logger-cf98bd5127572178.rlib --extern failure=/home/dd/work/ast_store/target/debug/deps/libfailure-da7463ce07672761.rlib --extern futures=/home/dd/work/ast_store/target/debug/deps/libfutures-fe721e4f842dc564.rlib --extern hex=/home/dd/work/ast_store/target/debug/deps/libhex-801f14eedb60222c.rlib --extern jsonwebtoken=/home/dd/work/ast_store/target/debug/deps/libjsonwebtoken-00f2e8867e3b1c8f.rlib --extern listenfd=/home/dd/work/ast_store/target/debug/deps/liblistenfd-b75707ba07f921d4.rlib --extern log=/home/dd/work/ast_store/target/debug/deps/liblog-fe56020e31c924f7.rlib --extern r2d2=/home/dd/work/ast_store/target/debug/deps/libr2d2-73c739306789f716.rlib --extern rand=/home/dd/work/ast_store/target/debug/deps/librand-e06e57be902b350f.rlib --extern rayon=/home/dd/work/ast_store/target/debug/deps/librayon-b70f929acdc3b0d5.rlib --extern crypto=/home/dd/work/ast_store/target/debug/deps/libcrypto-84a7de1b9fd676db.rlib --extern serde=/home/dd/work/ast_store/target/debug/deps/libserde-3127abc1b61429fa.rlib --extern serde_cbor=/home/dd/work/ast_store/target/debug/deps/libserde_cbor-edda23d0240421b5.rlib --extern serde_derive=/home/dd/work/ast_store/target/debug/deps/libserde_derive-efef5740544450c3.so --extern serde_json=/home/dd/work/ast_store/target/debug/deps/libserde_json-c1cd94b393a59f47.rlib --extern tempfile=/home/dd/work/ast_store/target/debug/deps/libtempfile-ba69c403ad632a75.rlib --extern tokio=/home/dd/work/ast_store/target/debug/deps/libtokio-cdda36ea935a2ce3.rlib --extern uuid=/home/dd/work/ast_store/target/debug/deps/libuuid-476d7cf0d1a23342.rlib --extern walkdir=/home/dd/work/ast_store/target/debug/deps/libwalkdir-4921a8c401aa6fcf.rlib -L native=/home/dd/work/ast_store/target/debug/build/backtrace-sys-ab2de583a5a63dbe/out -L native=/home/dd/work/ast_store/target/debug/build/brotli-sys-95a10507b4d53fe6/out -L native=/home/dd/work/ast_store/target/debug/build/ring-aa181c5e8c6d15dd/out -L native=/home/dd/work/ast_store/target/debug/build/miniz-sys-cb9ad76ca80f6a05/out -L native=/home/dd/work/ast_store/target/debug/build/asc-container-12a918cca80da199/out -L native=/home/dd/work/ast_store/target/debug/build/rust-crypto-f98b19eda631acd4/out -L native=/usr/lib/x86_64-linux-gnu -L native=/home/dd/work/ast_store/target/debug/build/libsqlite3-sys-14ba8e47f5445efc/out`
error[E0308]: mismatched types
   --> backend/src/lib.rs:232:86
    |
232 |         api::Request::ProductPriceList { version } => store::product_price_list(app, *version),
    |                                                                                      ^^^^^^^^ expected i64, found enum `std::option::Option`
    |
    = note: expected type `i64`
               found type `std::option::Option<i64>`

error[E0308]: mismatched types
   --> backend/src/lib.rs:232:86
    |
232 |         api::Request::ProductPriceList { version } => store::product_price_list(app, *version),
    |                                                                                      ^^^^^^^^ expected i64, found enum `std::option::Option`
    |
    = note: expected type `i64`
               found type `std::option::Option<i64>`

error[E0308]: mismatched types
   --> backend/src/lib.rs:232:86
    |
232 |         api::Request::ProductPriceList { version } => store::product_price_list(app, *version),
    |                                                                                      ^^^^^^^^ expected i64, found enum `std::option::Option`
    |
    = note: expected type `i64`
               found type `std::option::Option<i64>`

error: aborting due to previous error

For more information about this error, try `rustc --explain E0308`.
error: Could not compile `ast-backend`.

Caused by:
  process didn't exit successfully: `rustc --edition=2018 --crate-name ast_backend backend/src/lib.rs --color always --crate-type lib --emit=dep-info,link -C opt-level=1 -C panic=unwind -C debuginfo=2 -C debug-assertions=on -C metadata=235be9f8c403ef5f -C extra-filename=-235be9f8c403ef5f --out-dir /home/dd/work/ast_store/target/debug/deps -C incremental=/home/dd/work/ast_store/target/debug/incremental -L dependency=/home/dd/work/ast_store/target/debug/deps --extern actix=/home/dd/work/ast_store/target/debug/deps/libactix-e6e6ec5989c58f08.rlib --extern actix_web=/home/dd/work/ast_store/target/debug/deps/libactix_web-0c918bea4ce6e184.rlib --extern asc_container=/home/dd/work/ast_store/target/debug/deps/libasc_container-79248c028c0abaa6.rlib --extern asfile=/home/dd/work/ast_store/target/debug/deps/libasfile-501d87742c18de63.rlib --extern ast_store=/home/dd/work/ast_store/target/debug/deps/libast_store-e563fff47db84ca0.rlib --extern astindex2=/home/dd/work/ast_store/target/debug/deps/libastindex2-645dc7c64d98dc4b.rlib --extern astu_fw=/home/dd/work/ast_store/target/debug/deps/libastu_fw-eedd0a1b56692ddc.rlib --extern base64=/home/dd/work/ast_store/target/debug/deps/libbase64-e6a7e4ff2fe730df.rlib --extern bytes=/home/dd/work/ast_store/target/debug/deps/libbytes-e8b41f704e29044f.rlib --extern chrono=/home/dd/work/ast_store/target/debug/deps/libchrono-c9d85afdd546881d.rlib --extern diesel=/home/dd/work/ast_store/target/debug/deps/libdiesel-a097740f167c8b3a.rlib --extern dotenv=/home/dd/work/ast_store/target/debug/deps/libdotenv-7f00021dad648ea9.rlib --extern env_logger=/home/dd/work/ast_store/target/debug/deps/libenv_logger-97aaeafa161d4e8c.rlib --extern failure=/home/dd/work/ast_store/target/debug/deps/libfailure-e4fe55d2119b08e5.rlib --extern futures=/home/dd/work/ast_store/target/debug/deps/libfutures-692e72e42ffcb5bf.rlib --extern hex=/home/dd/work/ast_store/target/debug/deps/libhex-b793ef693e61c776.rlib --extern jsonwebtoken=/home/dd/work/ast_store/target/debug/deps/libjsonwebtoken-209fe79e69c28919.rlib --extern listenfd=/home/dd/work/ast_store/target/debug/deps/liblistenfd-c51dae87f5f318a2.rlib --extern log=/home/dd/work/ast_store/target/debug/deps/liblog-01c20663ef7f3d1f.rlib --extern r2d2=/home/dd/work/ast_store/target/debug/deps/libr2d2-0028d1814f157a08.rlib --extern rand=/home/dd/work/ast_store/target/debug/deps/librand-a37ae1bb175fe8d8.rlib --extern crypto=/home/dd/work/ast_store/target/debug/deps/libcrypto-4ef6247b79c391bc.rlib --extern serde=/home/dd/work/ast_store/target/debug/deps/libserde-aeb0ce1d2c705269.rlib --extern serde_cbor=/home/dd/work/ast_store/target/debug/deps/libserde_cbor-466e451a04b77cb2.rlib --extern serde_derive=/home/dd/work/ast_store/target/debug/deps/libserde_derive-efef5740544450c3.so --extern serde_json=/home/dd/work/ast_store/target/debug/deps/libserde_json-0454996f9a0511c1.rlib --extern tempfile=/home/dd/work/ast_store/target/debug/deps/libtempfile-a5cc25e0c0860958.rlib --extern tokio=/home/dd/work/ast_store/target/debug/deps/libtokio-1eecfad9964ed632.rlib --extern uuid=/home/dd/work/ast_store/target/debug/deps/libuuid-5d3f8589b71f0f58.rlib -L native=/home/dd/work/ast_store/target/debug/build/backtrace-sys-ab2de583a5a63dbe/out -L native=/home/dd/work/ast_store/target/debug/build/brotli-sys-95a10507b4d53fe6/out -L native=/home/dd/work/ast_store/target/debug/build/ring-aa181c5e8c6d15dd/out -L native=/home/dd/work/ast_store/target/debug/build/miniz-sys-cb9ad76ca80f6a05/out -L native=/home/dd/work/ast_store/target/debug/build/asc-container-12a918cca80da199/out -L native=/home/dd/work/ast_store/target/debug/build/rust-crypto-f98b19eda631acd4/out -L native=/usr/lib/x86_64-linux-gnu -L native=/home/dd/work/ast_store/target/debug/build/libsqlite3-sys-14ba8e47f5445efc/out` (exit code: 1)
warning: build failed, waiting for other jobs to finish...
error: aborting due to previous error

For more information about this error, try `rustc --explain E0308`.
error: aborting due to previous error

For more information about this error, try `rustc --explain E0308`.
error: Could not compile `ast-backend`.

Caused by:
  process didn't exit successfully: `rustc --edition=2018 --crate-name ast_backend backend/src/lib.rs --color always --emit=dep-info,link -C opt-level=1 -C debuginfo=2 -C debug-assertions=on --test -C metadata=b552aafba69820d4 -C extra-filename=-b552aafba69820d4 --out-dir /home/dd/work/ast_store/target/debug/deps -C incremental=/home/dd/work/ast_store/target/debug/incremental -L dependency=/home/dd/work/ast_store/target/debug/deps --extern actix=/home/dd/work/ast_store/target/debug/deps/libactix-33f6a9c015b4c03c.rlib --extern actix_web=/home/dd/work/ast_store/target/debug/deps/libactix_web-a11a56b0a62f4234.rlib --extern asc_container=/home/dd/work/ast_store/target/debug/deps/libasc_container-06f10de12fae04ac.rlib --extern asfile=/home/dd/work/ast_store/target/debug/deps/libasfile-4547961b3775e29e.rlib --extern ast_store=/home/dd/work/ast_store/target/debug/deps/libast_store-c5777bddc6b7a1ab.rlib --extern astindex2=/home/dd/work/ast_store/target/debug/deps/libastindex2-d64d1b076e5c74e6.rlib --extern astu_fw=/home/dd/work/ast_store/target/debug/deps/libastu_fw-43532442dc812bda.rlib --extern base64=/home/dd/work/ast_store/target/debug/deps/libbase64-f864812a9525c302.rlib --extern bytes=/home/dd/work/ast_store/target/debug/deps/libbytes-2e777982cd476fc8.rlib --extern chrono=/home/dd/work/ast_store/target/debug/deps/libchrono-31d6e1ac99b55645.rlib --extern diesel=/home/dd/work/ast_store/target/debug/deps/libdiesel-5eae397717655771.rlib --extern dotenv=/home/dd/work/ast_store/target/debug/deps/libdotenv-1fd097d50a0cbea9.rlib --extern env_logger=/home/dd/work/ast_store/target/debug/deps/libenv_logger-cf98bd5127572178.rlib --extern failure=/home/dd/work/ast_store/target/debug/deps/libfailure-da7463ce07672761.rlib --extern futures=/home/dd/work/ast_store/target/debug/deps/libfutures-fe721e4f842dc564.rlib --extern hex=/home/dd/work/ast_store/target/debug/deps/libhex-801f14eedb60222c.rlib --extern jsonwebtoken=/home/dd/work/ast_store/target/debug/deps/libjsonwebtoken-00f2e8867e3b1c8f.rlib --extern listenfd=/home/dd/work/ast_store/target/debug/deps/liblistenfd-b75707ba07f921d4.rlib --extern log=/home/dd/work/ast_store/target/debug/deps/liblog-fe56020e31c924f7.rlib --extern r2d2=/home/dd/work/ast_store/target/debug/deps/libr2d2-73c739306789f716.rlib --extern rand=/home/dd/work/ast_store/target/debug/deps/librand-e06e57be902b350f.rlib --extern rayon=/home/dd/work/ast_store/target/debug/deps/librayon-b70f929acdc3b0d5.rlib --extern crypto=/home/dd/work/ast_store/target/debug/deps/libcrypto-84a7de1b9fd676db.rlib --extern serde=/home/dd/work/ast_store/target/debug/deps/libserde-3127abc1b61429fa.rlib --extern serde_cbor=/home/dd/work/ast_store/target/debug/deps/libserde_cbor-edda23d0240421b5.rlib --extern serde_derive=/home/dd/work/ast_store/target/debug/deps/libserde_derive-efef5740544450c3.so --extern serde_json=/home/dd/work/ast_store/target/debug/deps/libserde_json-c1cd94b393a59f47.rlib --extern tempfile=/home/dd/work/ast_store/target/debug/deps/libtempfile-ba69c403ad632a75.rlib --extern tokio=/home/dd/work/ast_store/target/debug/deps/libtokio-cdda36ea935a2ce3.rlib --extern uuid=/home/dd/work/ast_store/target/debug/deps/libuuid-476d7cf0d1a23342.rlib --extern walkdir=/home/dd/work/ast_store/target/debug/deps/libwalkdir-4921a8c401aa6fcf.rlib -L native=/home/dd/work/ast_store/target/debug/build/backtrace-sys-ab2de583a5a63dbe/out -L native=/home/dd/work/ast_store/target/debug/build/brotli-sys-95a10507b4d53fe6/out -L native=/home/dd/work/ast_store/target/debug/build/ring-aa181c5e8c6d15dd/out -L native=/home/dd/work/ast_store/target/debug/build/miniz-sys-cb9ad76ca80f6a05/out -L native=/home/dd/work/ast_store/target/debug/build/asc-container-12a918cca80da199/out -L native=/home/dd/work/ast_store/target/debug/build/rust-crypto-f98b19eda631acd4/out -L native=/usr/lib/x86_64-linux-gnu -L native=/home/dd/work/ast_store/target/debug/build/libsqlite3-sys-14ba8e47f5445efc/out` (exit code: 1)
warning: build failed, waiting for other jobs to finish...
error: Could not compile `ast-backend`.

Caused by:
  process didn't exit successfully: `rustc --edition=2018 --crate-name ast_backend backend/src/lib.rs --color always --crate-type lib --emit=dep-info,link -C opt-level=1 -C debuginfo=2 -C debug-assertions=on -C metadata=f83015185db9fbc0 -C extra-filename=-f83015185db9fbc0 --out-dir /home/dd/work/ast_store/target/debug/deps -C incremental=/home/dd/work/ast_store/target/debug/incremental -L dependency=/home/dd/work/ast_store/target/debug/deps --extern actix=/home/dd/work/ast_store/target/debug/deps/libactix-33f6a9c015b4c03c.rlib --extern actix_web=/home/dd/work/ast_store/target/debug/deps/libactix_web-a11a56b0a62f4234.rlib --extern asc_container=/home/dd/work/ast_store/target/debug/deps/libasc_container-06f10de12fae04ac.rlib --extern asfile=/home/dd/work/ast_store/target/debug/deps/libasfile-4547961b3775e29e.rlib --extern ast_store=/home/dd/work/ast_store/target/debug/deps/libast_store-c5777bddc6b7a1ab.rlib --extern astindex2=/home/dd/work/ast_store/target/debug/deps/libastindex2-d64d1b076e5c74e6.rlib --extern astu_fw=/home/dd/work/ast_store/target/debug/deps/libastu_fw-43532442dc812bda.rlib --extern base64=/home/dd/work/ast_store/target/debug/deps/libbase64-f864812a9525c302.rlib --extern bytes=/home/dd/work/ast_store/target/debug/deps/libbytes-2e777982cd476fc8.rlib --extern chrono=/home/dd/work/ast_store/target/debug/deps/libchrono-31d6e1ac99b55645.rlib --extern diesel=/home/dd/work/ast_store/target/debug/deps/libdiesel-5eae397717655771.rlib --extern dotenv=/home/dd/work/ast_store/target/debug/deps/libdotenv-1fd097d50a0cbea9.rlib --extern env_logger=/home/dd/work/ast_store/target/debug/deps/libenv_logger-cf98bd5127572178.rlib --extern failure=/home/dd/work/ast_store/target/debug/deps/libfailure-da7463ce07672761.rlib --extern futures=/home/dd/work/ast_store/target/debug/deps/libfutures-fe721e4f842dc564.rlib --extern hex=/home/dd/work/ast_store/target/debug/deps/libhex-801f14eedb60222c.rlib --extern jsonwebtoken=/home/dd/work/ast_store/target/debug/deps/libjsonwebtoken-00f2e8867e3b1c8f.rlib --extern listenfd=/home/dd/work/ast_store/target/debug/deps/liblistenfd-b75707ba07f921d4.rlib --extern log=/home/dd/work/ast_store/target/debug/deps/liblog-fe56020e31c924f7.rlib --extern r2d2=/home/dd/work/ast_store/target/debug/deps/libr2d2-73c739306789f716.rlib --extern rand=/home/dd/work/ast_store/target/debug/deps/librand-e06e57be902b350f.rlib --extern crypto=/home/dd/work/ast_store/target/debug/deps/libcrypto-84a7de1b9fd676db.rlib --extern serde=/home/dd/work/ast_store/target/debug/deps/libserde-3127abc1b61429fa.rlib --extern serde_cbor=/home/dd/work/ast_store/target/debug/deps/libserde_cbor-edda23d0240421b5.rlib --extern serde_derive=/home/dd/work/ast_store/target/debug/deps/libserde_derive-efef5740544450c3.so --extern serde_json=/home/dd/work/ast_store/target/debug/deps/libserde_json-c1cd94b393a59f47.rlib --extern tempfile=/home/dd/work/ast_store/target/debug/deps/libtempfile-ba69c403ad632a75.rlib --extern tokio=/home/dd/work/ast_store/target/debug/deps/libtokio-cdda36ea935a2ce3.rlib --extern uuid=/home/dd/work/ast_store/target/debug/deps/libuuid-476d7cf0d1a23342.rlib -L native=/home/dd/work/ast_store/target/debug/build/backtrace-sys-ab2de583a5a63dbe/out -L native=/home/dd/work/ast_store/target/debug/build/brotli-sys-95a10507b4d53fe6/out -L native=/home/dd/work/ast_store/target/debug/build/ring-aa181c5e8c6d15dd/out -L native=/home/dd/work/ast_store/target/debug/build/miniz-sys-cb9ad76ca80f6a05/out -L native=/home/dd/work/ast_store/target/debug/build/asc-container-12a918cca80da199/out -L native=/home/dd/work/ast_store/target/debug/build/rust-crypto-f98b19eda631acd4/out -L native=/usr/lib/x86_64-linux-gnu -L native=/home/dd/work/ast_store/target/debug/build/libsqlite3-sys-14ba8e47f5445efc/out` (exit code: 1)
```

--- a/src/cargo/core/compiler/mod.rs
+++ b/src/cargo/core/compiler/mod.rs
@@ -23,7 +23,7 @@ use same_file::is_same_file;
 use serde::Serialize;
 
 use crate::core::manifest::TargetSourcePath;
-use crate::core::profiles::{Lto, Profile};
+use crate::core::profiles::{Lto, PanicStrategy, Profile};
 use crate::core::{PackageId, Target};
 use crate::util::errors::{CargoResult, CargoResultExt, Internal, ProcessError};
 use crate::util::paths;
@@ -794,7 +794,7 @@ fn build_base_args<'a, 'cfg>(
         cmd.arg("-C").arg(&format!("opt-level={}", opt_level));
     }
 
-    if let Some(panic) = panic.as_ref() {
+    if *panic != PanicStrategy::Unwind {
         cmd.arg("-C").arg(format!("panic={}", panic));
     }
 

--- a/src/cargo/core/profiles.rs
+++ b/src/cargo/core/profiles.rs
@@ -111,8 +111,8 @@ impl Profiles {
         let mut profile = maker.get_profile(Some(pkg_id), is_member, unit_for);
         // `panic` should not be set for tests/benches, or any of their
         // dependencies.
-        if !unit_for.is_panic_ok() || mode.is_any_test() {
-            profile.panic = None;
+        if !unit_for.is_panic_abort_ok() || mode.is_any_test() {
+            profile.panic = PanicStrategy::Unwind;
         }
 
         // Incremental can be globally overridden.
@@ -390,8 +390,13 @@ fn merge_profile(profile: &mut Profile, toml: &TomlProfile) {
     if let Some(rpath) = toml.rpath {
         profile.rpath = rpath;
     }
-    if let Some(ref panic) = toml.panic {
-        profile.panic = Some(InternedString::new(panic));
+    if let Some(panic) = &toml.panic {
+        profile.panic = match panic.as_str() {
+            "unwind" => PanicStrategy::Unwind,
+            "abort" => PanicStrategy::Abort,
+            // This should be validated in TomlProfile::validate
+            _ => panic!("Unexpected panic setting `{}`", panic),
+        };
     }
     if let Some(overflow_checks) = toml.overflow_checks {
         profile.overflow_checks = overflow_checks;
@@ -415,7 +420,7 @@ pub struct Profile {
     pub overflow_checks: bool,
     pub rpath: bool,
     pub incremental: bool,
-    pub panic: Option<InternedString>,
+    pub panic: PanicStrategy,
 }
 
 impl Default for Profile {
@@ -430,7 +435,7 @@ impl Default for Profile {
             overflow_checks: false,
             rpath: false,
             incremental: false,
-            panic: None,
+            panic: PanicStrategy::Unwind,
         }
     }
 }
@@ -530,26 +535,26 @@ impl Profile {
     fn comparable(
         &self,
     ) -> (
-        &InternedString,
-        &Lto,
-        &Option<u32>,
-        &Option<u32>,
-        &bool,
-        &bool,
-        &bool,
-        &bool,
-        &Option<InternedString>,
+        InternedString,
+        Lto,
+        Option<u32>,
+        Option<u32>,
+        bool,
+        bool,
+        bool,
+        bool,
+        PanicStrategy,
     ) {
         (
-            &self.opt_level,
-            &self.lto,
-            &self.codegen_units,
-            &self.debuginfo,
-            &self.debug_assertions,
-            &self.overflow_checks,
-            &self.rpath,
-            &self.incremental,
-            &self.panic,
+            self.opt_level,
+            self.lto,
+            self.codegen_units,
+            self.debuginfo,
+            self.debug_assertions,
+            self.overflow_checks,
+            self.rpath,
+            self.incremental,
+            self.panic,
         )
     }
 }
@@ -564,6 +569,23 @@ pub enum Lto {
     Named(InternedString),
 }
 
+/// The `panic` setting.
+#[derive(Clone, Copy, PartialEq, Eq, Debug, Hash, PartialOrd, Ord)]
+pub enum PanicStrategy {
+    Unwind,
+    Abort,
+}
+
+impl fmt::Display for PanicStrategy {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        match *self {
+            PanicStrategy::Unwind => "unwind",
+            PanicStrategy::Abort => "abort",
+        }
+        .fmt(f)
+    }
+}
+
 /// Flags used in creating `Unit`s to indicate the purpose for the target, and
 /// to ensure the target's dependencies have the correct settings.
 #[derive(Copy, Clone, Debug, Eq, PartialEq, Hash)]
@@ -571,11 +593,11 @@ pub struct UnitFor {
     /// A target for `build.rs` or any of its dependencies. This enables
     /// `build-override` profiles for these targets.
     custom_build: bool,
-    /// This is true if it is *allowed* to set the `panic` flag. Currently
+    /// This is true if it is *allowed* to set the `panic=abort` flag. Currently
     /// this is false for test/bench targets and all their dependencies, and
     /// "for_host" units such as proc macro and custom build scripts and their
     /// dependencies.
-    panic_ok: bool,
+    panic_abort_ok: bool,
 }
 
 impl UnitFor {
@@ -584,7 +606,7 @@ impl UnitFor {
     pub fn new_normal() -> UnitFor {
         UnitFor {
             custom_build: false,
-            panic_ok: true,
+            panic_abort_ok: true,
         }
     }
 
@@ -592,7 +614,7 @@ impl UnitFor {
     pub fn new_build() -> UnitFor {
         UnitFor {
             custom_build: true,
-            panic_ok: false,
+            panic_abort_ok: false,
         }
     }
 
@@ -600,7 +622,7 @@ impl UnitFor {
     pub fn new_compiler() -> UnitFor {
         UnitFor {
             custom_build: false,
-            panic_ok: false,
+            panic_abort_ok: false,
         }
     }
 
@@ -608,18 +630,18 @@ impl UnitFor {
     pub fn new_test() -> UnitFor {
         UnitFor {
             custom_build: false,
-            panic_ok: false,
+            panic_abort_ok: false,
         }
     }
 
     /// Creates a variant based on `for_host` setting.
     ///
-    /// When `for_host` is true, this clears `panic_ok` in a sticky fashion so
-    /// that all its dependencies also have `panic_ok=false`.
+    /// When `for_host` is true, this clears `panic_abort_ok` in a sticky fashion so
+    /// that all its dependencies also have `panic_abort_ok=false`.
     pub fn with_for_host(self, for_host: bool) -> UnitFor {
         UnitFor {
             custom_build: self.custom_build,
-            panic_ok: self.panic_ok && !for_host,
+            panic_abort_ok: self.panic_abort_ok && !for_host,
         }
     }
 
@@ -630,8 +652,8 @@ impl UnitFor {
     }
 
     /// Returns `true` if this unit is allowed to set the `panic` compiler flag.
-    pub fn is_panic_ok(self) -> bool {
-        self.panic_ok
+    pub fn is_panic_abort_ok(self) -> bool {
+        self.panic_abort_ok
     }
 
     /// All possible values, used by `clean`.
@@ -639,15 +661,15 @@ impl UnitFor {
         static ALL: [UnitFor; 3] = [
             UnitFor {
                 custom_build: false,
-                panic_ok: true,
+                panic_abort_ok: true,
             },
             UnitFor {
                 custom_build: true,
-                panic_ok: false,
+                panic_abort_ok: false,
             },
             UnitFor {
                 custom_build: false,
-                panic_ok: false,
+                panic_abort_ok: false,
             },
         ];
         &ALL

--- a/src/cargo/util/toml/mod.rs
+++ b/src/cargo/util/toml/mod.rs
@@ -5,6 +5,7 @@ use std::path::{Path, PathBuf};
 use std::rc::Rc;
 use std::str;
 
+use failure::bail;
 use log::{debug, trace};
 use semver::{self, VersionReq};
 use serde::de;
@@ -52,8 +53,9 @@ fn do_read_manifest(
     let package_root = manifest_file.parent().unwrap();
 
     let toml = {
-        let pretty_filename =
-            manifest_file.strip_prefix(config.cwd()).unwrap_or(manifest_file);
+        let pretty_filename = manifest_file
+            .strip_prefix(config.cwd())
+            .unwrap_or(manifest_file);
         parse(contents, pretty_filename, config)?
     };
 
@@ -78,7 +80,7 @@ fn do_read_manifest(
             TomlManifest::to_real_manifest(&manifest, source_id, package_root, config)?;
         add_unused(manifest.warnings_mut());
         if !manifest.targets().iter().any(|t| !t.is_custom_build()) {
-            failure::bail!(
+            bail!(
                 "no targets specified in the manifest\n  \
                  either src/lib.rs, src/main.rs, a [lib] section, or \
                  [[bin]] section must be present"
@@ -471,7 +473,7 @@ impl TomlProfile {
             "dev" | "release" => {}
             _ => {
                 if self.overrides.is_some() || self.build_override.is_some() {
-                    failure::bail!(
+                    bail!(
                         "Profile overrides may only be specified for \
                          `dev` or `release` profile, not `{}`.",
                         name
@@ -491,21 +493,31 @@ impl TomlProfile {
             }
             _ => {}
         }
+
+        if let Some(panic) = &self.panic {
+            if panic != "unwind" && panic != "abort" {
+                bail!(
+                    "`panic` setting of `{}` is not a valid setting,\
+                     must be `unwind` or `abort`",
+                    panic
+                );
+            }
+        }
         Ok(())
     }
 
     fn validate_override(&self) -> CargoResult<()> {
         if self.overrides.is_some() || self.build_override.is_some() {
-            failure::bail!("Profile overrides cannot be nested.");
+            bail!("Profile overrides cannot be nested.");
         }
         if self.panic.is_some() {
-            failure::bail!("`panic` may not be specified in a profile override.")
+            bail!("`panic` may not be specified in a profile override.")
         }
         if self.lto.is_some() {
-            failure::bail!("`lto` may not be specified in a profile override.")
+            bail!("`lto` may not be specified in a profile override.")
         }
         if self.rpath.is_some() {
-            failure::bail!("`rpath` may not be specified in a profile override.")
+            bail!("`rpath` may not be specified in a profile override.")
         }
         Ok(())
     }
@@ -830,7 +842,7 @@ impl TomlManifest {
 
         let package_name = project.name.trim();
         if package_name.is_empty() {
-            failure::bail!("package name cannot be an empty string")
+            bail!("package name cannot be an empty string")
         }
 
         validate_package_name(package_name, "package name", "")?;
@@ -951,7 +963,7 @@ impl TomlManifest {
                 let name = dep.name_in_toml();
                 let prev = names_sources.insert(name.to_string(), dep.source_id());
                 if prev.is_some() && prev != Some(dep.source_id()) {
-                    failure::bail!(
+                    bail!(
                         "Dependency '{}' has different source paths depending on the build \
                          target. Each dependency must have a single canonical source path \
                          irrespective of build target.",
@@ -1006,7 +1018,7 @@ impl TomlManifest {
             (None, root) => WorkspaceConfig::Member {
                 root: root.cloned(),
             },
-            (Some(..), Some(..)) => failure::bail!(
+            (Some(..), Some(..)) => bail!(
                 "cannot configure both `package.workspace` and \
                  `[workspace]`, only one can be specified"
             ),
@@ -1082,43 +1094,43 @@ impl TomlManifest {
         config: &Config,
     ) -> CargoResult<(VirtualManifest, Vec<PathBuf>)> {
         if me.project.is_some() {
-            failure::bail!("virtual manifests do not define [project]");
+            bail!("virtual manifests do not define [project]");
         }
         if me.package.is_some() {
-            failure::bail!("virtual manifests do not define [package]");
+            bail!("virtual manifests do not define [package]");
         }
         if me.lib.is_some() {
-            failure::bail!("virtual manifests do not specify [lib]");
+            bail!("virtual manifests do not specify [lib]");
         }
         if me.bin.is_some() {
-            failure::bail!("virtual manifests do not specify [[bin]]");
+            bail!("virtual manifests do not specify [[bin]]");
         }
         if me.example.is_some() {
-            failure::bail!("virtual manifests do not specify [[example]]");
+            bail!("virtual manifests do not specify [[example]]");
         }
         if me.test.is_some() {
-            failure::bail!("virtual manifests do not specify [[test]]");
+            bail!("virtual manifests do not specify [[test]]");
         }
         if me.bench.is_some() {
-            failure::bail!("virtual manifests do not specify [[bench]]");
+            bail!("virtual manifests do not specify [[bench]]");
         }
         if me.dependencies.is_some() {
-            failure::bail!("virtual manifests do not specify [dependencies]");
+            bail!("virtual manifests do not specify [dependencies]");
         }
         if me.dev_dependencies.is_some() || me.dev_dependencies2.is_some() {
-            failure::bail!("virtual manifests do not specify [dev-dependencies]");
+            bail!("virtual manifests do not specify [dev-dependencies]");
         }
         if me.build_dependencies.is_some() || me.build_dependencies2.is_some() {
-            failure::bail!("virtual manifests do not specify [build-dependencies]");
+            bail!("virtual manifests do not specify [build-dependencies]");
         }
         if me.features.is_some() {
-            failure::bail!("virtual manifests do not specify [features]");
+            bail!("virtual manifests do not specify [features]");
         }
         if me.target.is_some() {
-            failure::bail!("virtual manifests do not specify [target]");
+            bail!("virtual manifests do not specify [target]");
         }
         if me.badges.is_some() {
-            failure::bail!("virtual manifests do not specify [badges]");
+            bail!("virtual manifests do not specify [badges]");
         }
 
         let mut nested_paths = Vec::new();
@@ -1151,7 +1163,7 @@ impl TomlManifest {
                 &config.exclude,
             )),
             None => {
-                failure::bail!("virtual manifests must be configured with [workspace]");
+                bail!("virtual manifests must be configured with [workspace]");
             }
         };
         Ok((
@@ -1162,7 +1174,7 @@ impl TomlManifest {
 
     fn replace(&self, cx: &mut Context<'_, '_>) -> CargoResult<Vec<(PackageIdSpec, Dependency)>> {
         if self.patch.is_some() && self.replace.is_some() {
-            failure::bail!("cannot specify both [replace] and [patch]");
+            bail!("cannot specify both [replace] and [patch]");
         }
         let mut replace = Vec::new();
         for (spec, replacement) in self.replace.iter().flat_map(|x| x) {
@@ -1182,7 +1194,7 @@ impl TomlManifest {
                 TomlDependency::Simple(..) => true,
             };
             if version_specified {
-                failure::bail!(
+                bail!(
                     "replacements cannot specify a version \
                      requirement, but found one for `{}`",
                     spec
@@ -1331,12 +1343,12 @@ impl DetailedTomlDependency {
             self.registry.as_ref(),
             self.registry_index.as_ref(),
         ) {
-            (Some(_), _, Some(_), _) | (Some(_), _, _, Some(_)) => failure::bail!(
+            (Some(_), _, Some(_), _) | (Some(_), _, _, Some(_)) => bail!(
                 "dependency ({}) specification is ambiguous. \
                  Only one of `git` or `registry` is allowed.",
                 name_in_toml
             ),
-            (_, _, Some(_), Some(_)) => failure::bail!(
+            (_, _, Some(_), Some(_)) => bail!(
                 "dependency ({}) specification is ambiguous. \
                  Only one of `registry` or `registry-index` is allowed.",
                 name_in_toml

--- a/tests/testsuite/profiles.rs
+++ b/tests/testsuite/profiles.rs
@@ -376,3 +376,40 @@ fn profile_doc_deprecated() {
         .with_stderr_contains("[WARNING] profile `doc` is deprecated and has no effect")
         .run();
 }
+
+#[test]
+fn panic_unwind_does_not_build_twice() {
+    // Check for a bug where `lib` was built twice, once with panic set and
+    // once without. Since "unwind" is the default, they are the same and
+    // should only be built once.
+    let p = project()
+        .file(
+            "Cargo.toml",
+            r#"
+            [package]
+            name = "foo"
+            version = "0.1.0"
+
+            [profile.dev]
+            panic = "unwind"
+            "#,
+        )
+        .file("src/lib.rs", "")
+        .file("src/main.rs", "fn main() {}")
+        .file("tests/t1.rs", "")
+        .build();
+
+    p.cargo("test -v --tests --no-run")
+        .with_stderr_unordered(
+            "\
+[COMPILING] foo [..]
+[RUNNING] `rustc --crate-name foo src/lib.rs [..]--crate-type lib [..]
+[RUNNING] `rustc --crate-name foo src/lib.rs [..] --test [..]
+[RUNNING] `rustc --crate-name foo src/main.rs [..]--crate-type bin [..]
+[RUNNING] `rustc --crate-name foo src/main.rs [..] --test [..]
+[RUNNING] `rustc --crate-name t1 tests/t1.rs [..]
+[FINISHED] [..]
+",
+        )
+        .run();
+}

